src/coreclr/jit/compiler.cpp

```diff
@@ -5298,100 +5479,36 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
         {
             // Loop alignment is disabled for cold blocks
             assert(!block->HasFlag(BBF_COLD));
-            BasicBlock* const loopTop              = block->Next();
-            bool              isSpecialCallFinally = block->isBBCallAlwaysPairTail();
-            bool              unmarkedLoopAlign    = false;
+            BasicBlock* const loopTop = block->Next();
 
-#if FEATURE_EH_CALLFINALLY_THUNKS
-            if (block->KindIs(BBJ_CALLFINALLY))
+            if (bbHavingAlign == nullptr)
             {
-                // It must be a retless BBJ_CALLFINALLY if we get here.
-                assert(!block->isBBCallAlwaysPair());
+                // If jmp was not found, then block before the loop start is where align instruction will be added.
 
-                // In the case of FEATURE_EH_CALLFINALLY_THUNKS, we can't put the align instruction in a retless
-                // BBJ_CALLFINALLY either, because it alters the "cloned finally" region reported to the VM.
-                // In the x86 case (the only !FEATURE_EH_CALLFINALLY_THUNKS that supports retless
-                // BBJ_CALLFINALLY), we allow it.
-                isSpecialCallFinally = true;
+                bbHavingAlign = block;
+                JITDUMP("Marking " FMT_BB " before the loop with BBF_HAS_ALIGN for loop at " FMT_BB "\n", block->bbNum,
+                        loopTop->bbNum);
             }
-#endif // FEATURE_EH_CALLFINALLY_THUNKS
-
-            if (isSpecialCallFinally)
+            else
             {
-                // There are two special cases:
-                // 1. If the block before the loop start is a retless BBJ_CALLFINALLY with
-                //    FEATURE_EH_CALLFINALLY_THUNKS, we can't add alignment because it will affect reported EH
-                //    region range.
-                // 2. If the previous block is the BBJ_ALWAYS of a BBJ_CALLFINALLY/BBJ_ALWAYS pair, then we
-                //    can't add alignment because we can't add instructions in that block. In the
-                //    FEATURE_EH_CALLFINALLY_THUNKS case, it would affect the reported EH, as above.
-                // Currently, we don't align loops for these cases.
-
-                loopTop->unmarkLoopAlign(this DEBUG_ARG("block before loop is special callfinally/always block"));
-                madeChanges       = true;
-                unmarkedLoopAlign = true;
+                JITDUMP("Marking " FMT_BB " that ends with unconditional jump with BBF_HAS_ALIGN for loop at " FMT_BB
+                        "\n",
+                        bbHavingAlign->bbNum, loopTop->bbNum);
             }
-            else if ((loopTop->bbNatLoopNum != BasicBlock::NOT_IN_LOOP) && visitedLoopNum[loopTop->bbNatLoopNum])
```

Comment:
It's not possible -- `optIdentifyLoopsForAlignment` will only add `BBF_LOOP_ALIGN` to the lexical top-most block of candidates for loop alignment. Asserting it is not so simple, but I expect with my follow-up cleanup it will look more obvious (we will have a bit set like before tracking loops we have seen)